### PR TITLE
Support for DIP1009 (new contracts syntax), #375

### PR DESCRIPTION
**Pull request by LaurentTreguier**
I don't know if what I did is a correct way to do it (I've never really touched lexing and such before).
In `isBlockHeaderToken()` I added `in` because just like `out`, it can be a block header with classic contract syntax (maybe that should be in a separate commit ?).
In `isBlockHeader()`, I think changing `i + index >= tokens.length` to `i + index + 1 >= tokens.length` should not be a problem since even if the token is not outside of the range of `tokens`, if it's the very last one it shouldn't be a block header (if it is, then we'd be formatting invalid code anyway).

### CODE REVIEW

**Comment by LaurentTreguier** [comment]
I was way too hasty and forgot to run the tests...

**Comment by BBasile** [comment]
You must add tests with the new syntax too.

**Comment by LaurentTreguier** [comment]
Yes. I guess I have to start thinking before acting...

**Comment by LaurentTreguier** [comment]
When I made this PR I had completely forgotten that obviously `out` blocks can have an argument too; so the only difference in this case is that block-based `out` contracts have a block following the argument...

So I have something that seems to be working (that doesn't break the tests, and with its own test files), however I had to make changes to libdparse again to resolve this problem: I added a `isExpression` property to the `InStatement` and `OutStatement` node types, since I didn't find a way to differentiate an expression-based contract from a block-based one in dfmt.

Do you see another way to do this ?

**Comment by BBasile** [comment]
`out` followed by a left paren is always a contract. Isn't this enough to format a one line expression ?

**Comment by LaurentTreguier** [comment]
`out` contracts followed by a left paren can be a block one : https://dlang.org/spec/contracts.html#pre_post_contracts

**Comment by BBasile** [comment]
actually (sorry previous answer was wrong) you should be able to detect an expression based contract with the peek function family, e.g

```d
 if (peeIs(tok!"(", true) &&
     // out(;...)
    (peek2Is(tok!";", true) ||
    // out(a;...)
    (peek2Is(tok!"identifier", true) && peek2Is(tok!";", true)))){}   
```

**Comment by BBasile** [comment]
edit: peek3Is

**Comment by LaurentTreguier** [comment]
And this time I forgot about to update the git submodule. When I can get back on this the next push shall fix it for good, I swear !!

File: src/dfmt/formatter.d [hunk 1]
```diff
@@ -1908,8 +1908,21 @@ const pure @safe @nogc:
...
+        const token = tokens[i + index];
+        const t = token.type;
+        bool isExpressionContract;
+
+        if (i + index + 3 < tokens.length)
+        {
+            IdType[3] toks;
+            foreach (j, ref tok; toks)
+                tok = tokens[i + index + j + 1].type;
+            isExpressionContract = (t == tok!"in" && toks[0] == tok!"(")
```
[the remark this answers is not in the document]
Comment (BBasile) [reply src/dfmt/formatter.d]: test case

```d
out /*Major*/ ( /*Tom*/ result /*To ground control*/; true)
```

**Comment by BBasile** [comment]
rebase against `master` please

**Comment by BBasile** [comment]
You can approve this one https://github.com/dlang-community/dfmt/pull/383, after what you'll have to put the auto merge label and pull, rebase again to make the CI status green

File: tests/allman/dip1009.d.ref [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    assert(result == 0);
+}
+do
+{
+    return 0;
+}
+
+int bar(int arg)
+in(arg > 0)
+out(; true);
```
Comment (BBasile) [inline tests/allman/dip1009.d.ref]: I missed that: trailing `;` is invalid.

Comment (BBasile) [reply tests/allman/dip1009.d.ref]: i don't know how the tests site can work with this error

Comment (LaurentTreguier) [reply tests/allman/dip1009.d.ref]: As far as I understand the tests for dfmt concentrate on the formatted output, and not on the correctness of the input; putting this in libdparse's tests does make its test suite fail

**Pull request merged**